### PR TITLE
Fixes to make LDX and STX conformance tests pass

### LINKED ISSUE #420: Some ldx*.data conformance tests fail verification
The following ldx conformance tests fail verification, which means valid programs might be rejected:

* ldxb
* ldxdw
* ldxh
* ldxw

### PR DESCRIPTION
**Pull request by dthaler**
Also make stack.data conformance test pass

Fixes #420

Signed-off-by: Dave Thaler <dthaler@microsoft.com>

### CODE REVIEW

**Review by elazarg — COMMENTED** [review]
Some style comments. I hope to get a deeper look soon. 

@caballa - it would be nice if you could have a look at this one too.

File: src/crab/array_domain.cpp [hunk 1]
```diff
@@ -508,6 +525,57 @@ std::optional<linear_expression_t> array_domain_t::load(NumAbsDomain& inv, data_
...
+                if (!b) {
+                    b = get_value_byte(inv, byte_offset, 4);
+                    if (!b) {
+                        b = get_value_byte(inv, byte_offset, 2);
+                        if (!b) {
+                            b = get_value_byte(inv, byte_offset, 1);
+                        }
+                    }
+                }
+                if (b) {
```
[the remark this answers is not in the document]
Comment (dthaler) [reply src/crab/array_domain.cpp:550]: No, `?:` is not standard C++, it is a GNU C extension that does not work with other compilers.
https://en.wikipedia.org/wiki/Ternary_conditional_operator#C_Variants explains
"A [GNU](https://en.wikipedia.org/wiki/GNU_Project) extension to C allows omitting the second operand ..."

File: src/crab/array_domain.cpp [hunk 2]
```diff
@@ -508,6 +525,57 @@ std::optional<linear_expression_t> array_domain_t::load(NumAbsDomain& inv, data_
         }
         offset_t o(k);
         unsigned size = (long)width;
+        std::optional<cell_t> cell = lookup_array_map(kind).get_cell(o, size);
+        if (cell) {
```
[the remark this answers is not in the document]
Comment (dthaler) [reply src/crab/array_domain.cpp]: Done

File: src/crab/array_domain.cpp [hunk 3]
```diff
@@ -508,6 +525,56 @@ std::optional<linear_expression_t> array_domain_t::load(NumAbsDomain& inv, data_
...
+            return cell->get_scalar(kind);
+        }
+        if (kind == data_kind_t::values) {
+            // Copy bytes into result_buffer, taking into account that the
+            // bytes might be in different stack variables and might be unaligned.
+            uint8_t result_buffer[8];
+            bool found = true;
+            for (unsigned int index = 0; index < size; index++) {
+                offset_t byte_offset = o + index;
+                uint64_t buffer;
```
Comment (elazarg) [inline src/crab/array_domain.cpp:538]:
```suggestion
```

**Review by elazarg — APPROVED** [review]
LGTM. It would be nice to extract the entire `if kind == data_kind_t::values` case to a function returning `optional<linear_expression_t>` since the method becomes quite large, and the code is not obvious.

I did not commit the deletion of the unused variable just in case it indicates something that was missed. I will merge after this is fixed.

**Comment by dthaler** [comment]
> I did not commit the deletion of the unused variable just in case it indicates something that was missed. I will merge after this is fixed.

Committed deletion, ok to merge now.

**Pull request merged**